diff --git a/target/i386/cpu-dump.c b/target/i386/cpu-dump.c
index 40697064d9..7ccd64e974 100644
--- a/target/i386/cpu-dump.c
+++ b/target/i386/cpu-dump.c
@@ -99,9 +99,9 @@ cpu_x86_dump_seg_cache(CPUX86State *env, FILE *f,
 {
 #ifdef TARGET_X86_64
     if (env->hflags & HF_CS64_MASK) {
-        qemu_fprintf(f, "%-3s=%04x %016" PRIx64 " %08x %08x", name,
+        qemu_fprintf(f, "%-3s=%04x %016" PRIx64 " %08x %x", name,
                      sc->selector, sc->base, sc->limit,
-                     sc->flags & 0x00ffff00);
+                     sc->flags);
     } else
 #endif
     {
@@ -421,6 +421,32 @@ void x86_cpu_dump_state(CPUState *cs, FILE *f, int flags)
     cpu_x86_dump_seg_cache(env, f, "LDT", &env->ldt);
     cpu_x86_dump_seg_cache(env, f, "TR", &env->tr);
 
+#define dumpmsr(name) \
+    qemu_fprintf(f, #name " = %lx\n", env-> name );
+#define dumpmsr32(name) \
+    qemu_fprintf(f, #name " = %x\n", env-> name );
+
+    dumpmsr(kernelgsbase);
+    dumpmsr(lstar);
+    dumpmsr(cstar);
+    dumpmsr(fmask);
+
+    dumpmsr32(sysenter_cs);
+    dumpmsr(sysenter_esp);
+    dumpmsr(sysenter_eip);
+    dumpmsr(star);
+
+    dumpmsr(tsc_deadline);
+    dumpmsr(tsc_aux);
+
+    dumpmsr(pat);
+    dumpmsr(xcr0);
+
+
+    uint64_t val = cpu_get_apic_base(env_archcpu(env)->apic_state);
+    qemu_fprintf(f, "apicbase =     %lx\n", val );
+
+
 #ifdef TARGET_X86_64
     if (env->hflags & HF_LMA_MASK) {
         qemu_fprintf(f, "GDT=     %016" PRIx64 " %08x\n",
diff --git a/target/i386/kvm/kvm.c b/target/i386/kvm/kvm.c
index de531842f6..95b0af6e51 100644
--- a/target/i386/kvm/kvm.c
+++ b/target/i386/kvm/kvm.c
@@ -5529,6 +5529,12 @@ int kvm_arch_handle_exit(CPUState *cs, struct kvm_run *run)
         ret = kvm_xen_handle_exit(cpu, &run->xen);
         break;
 #endif
+    case KVM_EXIT_HYPERCALL:
+        printf("=== HYPERPILL HYPERCALL RECEIVED ===\n");
+        kvm_arch_get_registers(cs);
+        vm_stop(RUN_STATE_PAUSED);
+        ret = 0;
+        break;
     default:
         fprintf(stderr, "KVM: unknown exit reason %d\n", run->exit_reason);
         ret = -1;
