From 3019fcc1c082cfa16cb20bd654fe5cd6018b4bb0 Mon Sep 17 00:00:00 2001
From: Qiang Liu <cyruscyliu@gmail.com>
Date: Fri, 9 Aug 2024 15:18:03 +0200
Subject: [PATCH] update OpenCore-Boot.sh to support vmx

---
 OpenCore-Boot.sh | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/OpenCore-Boot.sh b/OpenCore-Boot.sh
index e027851..f44b156 100755
--- a/OpenCore-Boot.sh
+++ b/OpenCore-Boot.sh
@@ -18,7 +18,7 @@
 ###############################################################################
 
 
-MY_OPTIONS="+ssse3,+sse4.2,+popcnt,+avx,+aes,+xsave,+xsaveopt,check"
+MY_OPTIONS="+vmx,+ssse3,+sse4.2,+popcnt,+avx,+aes,+xsave,+xsaveopt,check"
 
 # This script works for Big Sur, Catalina, Mojave, and High Sierra. Tested with
 # macOS 10.15.6, macOS 10.14.6, and macOS 10.13.6.
@@ -33,7 +33,7 @@ OVMF_DIR="."
 
 # shellcheck disable=SC2054
 args=(
-  -enable-kvm -m "$ALLOCATED_RAM" -cpu Penryn,kvm=on,vendor=GenuineIntel,+invtsc,vmware-cpuid-freq=on,"$MY_OPTIONS"
+  -enable-kvm -m "$ALLOCATED_RAM" -cpu Skylake-Client,kvm=on,vendor=GenuineIntel,+invtsc,vmware-cpuid-freq=on,"$MY_OPTIONS"
   -machine q35
   -device qemu-xhci,id=xhci
   -device usb-kbd,bus=xhci.0 -device usb-tablet,bus=xhci.0
-- 
2.34.1

